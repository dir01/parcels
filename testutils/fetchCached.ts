export async function fetchCached(
  url: string | URL | Request,
  init?: RequestInit | undefined,
): Promise<Response> {
  const force = Deno.env.get("FORCE_FETCH");
  if (force || !(await existsCached(url, init))) {
    const resp = await fetch(url, init);
    await cacheResponse(url, init, resp);
    return resp;
  }

  return await getCachedResponse(url, init);
}

function getFilename(
  url: string | URL | Request,
  init: RequestInit | undefined,
) {
  const urlStr = typeof url === "string" ? url : url.toString();
  const initStr = init ? JSON.stringify(init) : "";
  let filename = urlStr + "_" + initStr;
  filename = filename.replace(/[^a-zA-Z0-9]/g, "_");
  filename = filename.replace(/_$/g, "");
  return `./testdata/${filename}.json`;
}

async function existsCached(
  url: string | URL | Request,
  init: RequestInit | undefined,
) {
  const filename = getFilename(url, init);
  try {
    const stat = await Deno.stat(filename);
    return stat.isFile;
  } catch {
    return false;
  }
}

async function cacheResponse(
  url: string | URL | Request,
  init: RequestInit | undefined,
  resp: Response,
) {
  const filename = getFilename(url, init);
  await Deno.writeFile(filename, await serializeResponse(resp));
}

async function getCachedResponse(
  url: string | URL | Request,
  init: RequestInit | undefined,
) {
  const filename = getFilename(url, init);
  return deserializeResponse(await Deno.readFile(filename));
}

async function serializeResponse(resp: Response): Promise<Uint8Array> {
  let body = await resp.text();
  try {
    body = JSON.parse(body);
  } catch {
    // ignore
  }

  const s = JSON.stringify(
    {
      status: resp.status,
      statusText: resp.statusText,
      headers: [...resp.headers],
      body: body,
    },
    null,
    2,
  );
  return new TextEncoder().encode(s);
}

function deserializeResponse(resp: Uint8Array): Response {
  const s = new TextDecoder().decode(resp);
  const { status, statusText, headers, body } = JSON.parse(s);
  let serializedBody = body;
  if (typeof body === "object") {
    serializedBody = JSON.stringify(body, null, 2);
  }
  return new Response(serializedBody, {
    status,
    statusText,
    headers: new Headers(headers),
  });
}
