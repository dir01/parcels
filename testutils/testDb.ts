import { DB } from "https://deno.land/x/sqlite@v3.7.0/mod.ts";

export function prepareTestDb() {
  const db = new DB(":memory:");
  db.execute(
    `
    CREATE TABLE postal_api_responses (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      api_name TEXT NOT NULL,
      tracking_number TEXT NOT NULL,
      first_fetched_at INTEGER NOT NULL,
      last_fetched_at INTEGER NOT NULL,
      response_body TEXT NOT NULL,
      status TEXT NOT NULL,
      error TEXT
    );
    `,
  );
  return db;
}
