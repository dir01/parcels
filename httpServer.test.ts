import { assertEquals } from "https://deno.land/std@0.173.0/testing/asserts.ts";
import { assertSnapshot } from "https://deno.land/std@0.168.0/testing/snapshot.ts";
import { mockFetch, unMockFetch } from "https://deno.land/x/metch@0.1.0/mod.ts";
import HttpServer from "./httpServer.ts";
import PostalService from "./postalService.ts";
import SQLitePostalApiResponseStorage from "./storage.ts";
import CainiaoApi from "./apis/cainiao/cainiao.ts";
import { fetchCached } from "./testutils/fetchCached.ts";
import { prepareTestDb } from "./testutils/testDb.ts";

Deno.test("HTTP server", async (t) => {
  const postalService = new PostalService({
    storage: new SQLitePostalApiResponseStorage({ db: prepareTestDb() }),
    hoursBetweenChecks: 0,
    expiryDays: 0,
    postalApiMap: {
      cainiao: new CainiaoApi(),
    },
    now: () => new Date("2020-01-01T00:00:00.000Z"),
  });
  const server = new HttpServer({ postalService, port: 9000 });

  async function sendRequest(trackingNumber: string) {
    const req = new Request("http://whatever/", {
      method: "POST",
      body: JSON.stringify({ trackingNumber }),
    });
    return await server.handleRequest(req);
  }

  await t.step("cainiao", async () => {
    await withMockedFetch(
      "https://global.cainiao.com/global/detail.json?mailNos=AE010698498&lang=en-US",
      async () => {
        const resp = await sendRequest("AE010698498");
        const resBody = await resp.json();

        assertEquals(resp.status, 200);
        assertEquals(resBody.length, 1);
        assertEquals(resBody[0].trackingNumber, "AE010698498");
        await assertSnapshot(t, resBody);
      }
    );
  });
});

async function withMockedFetch(url: string, f: () => Promise<void>) {
  await mockFetch(new Request(url), await fetchCached(url));
  try {
    await f();
  } finally {
    unMockFetch();
  }
}
