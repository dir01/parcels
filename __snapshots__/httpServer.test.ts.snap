export const snapshot = {};

snapshot[`HTTP server 1`] = `
[
  {
    apiName: "cainiao",
    destinationCountry: "Israel",
    events: [
      {
        code: "IMPORT_CUSTOMS_CLEARANCE_SUCCESS",
        description: "Import customs clearance complete",
        time: "2023-01-29T00:31:00.000Z",
      },
      {
        code: "DEPARTED_FROM_CUSTOMS",
        description: "Departed from customs",
        time: "2023-01-29T00:31:00.000Z",
      },
      {
        code: "ARRIVED_AT_CUSTOMS",
        description: "Arrived at customs",
        time: "2023-01-28T21:00:00.000Z",
      },
      {
        code: "ARRIVED_AT_CUSTOMS",
        description: "Carrier update",
        time: "2023-01-28T21:00:00.000Z",
      },
      {
        code: "ARRIVED_AT_LINEHAUL_OFFICE",
        description: "Arrived at linehual office",
        time: "2023-01-27T05:23:00.000Z",
      },
      {
        code: "DEPARTED_ORIGIN_COUNTRY",
        description: "Departed from departure country/region",
        time: "2023-01-26T15:18:00.000Z",
      },
      {
        code: "IMPORT_CUSTOMS_CLEARANCE_STARTED",
        description: "Import customs clearance started",
        time: "2023-01-26T01:47:00.000Z",
      },
      {
        code: "LH_HO_AIRLINE",
        description: "Leaving from departure country/region",
        time: "2023-01-26T00:00:00.000Z",
      },
      {
        code: "EXPORT_CUSTOMS_CLEARANCE_SUCCESS",
        description: "Export customs clearance complete",
        time: "2023-01-25T06:55:42.000Z",
      },
      {
        code: "EXPORT_CUSTOMS_CLEARANCE_STARTED",
        description: "Export customs clearance started",
        time: "2023-01-25T06:35:19.000Z",
      },
      {
        code: "TRANSIT_PORT_REROUTE_CALLBACK",
        description: "Carrier update",
        time: "2023-01-25T03:50:38.471Z",
      },
      {
        code: "LH_HO_IN_SUCCESS",
        description: "Arrived at departure transport hub",
        time: "2023-01-25T00:25:15.000Z",
      },
      {
        code: "DEPARTED_FROM_SORTING_CENTER",
        description: "[Shatian Town] Departed from sorting center",
        time: "2023-01-24T12:54:35.000Z",
      },
      {
        code: "ACCEPTED_BY_CARRIER",
        description: "Received by logistics company",
        time: "2023-01-24T12:20:21.000Z",
      },
      {
        code: "ARRIVED_AT_SORTING_CENTER",
        description: "[Shatian Town] Processing at sorting center",
        time: "2023-01-24T12:20:21.000Z",
      },
      {
        code: "WMS_CONFIRMED",
        description: "Carrier update",
        time: "2023-01-24T03:30:03.862Z",
      },
      {
        code: "DISPATCHED_FROM_WAREHOUSE",
        description: "Package shipped out from warehouse",
        time: "2023-01-24T03:30:03.000Z",
      },
      {
        code: "PACKAGING_COMPLETED",
        description: "Package ready for shipping from warehouse",
        time: "2023-01-24T03:26:28.216Z",
      },
      {
        code: "SHIPMENT_INFO_RECEIVED",
        description: "Shipment information received by warehouse electronically",
        time: "2023-01-23T17:59:33.895Z",
      },
    ],
    fetchedAt: "2020-01-01T00:00:00.000Z",
    originCountry: "Mainland China",
    trackingNumber: "AE010698498",
  },
]
`;
